pystiche/papers/ulyanov_et_al_2016/loss.py

```diff
@@ -0,0 +1,167 @@
+    def __init__(
+        self, encoder: Encoder, impl_params: bool = True, **gram_op_kwargs: Any,
+    ):
+        super().__init__(encoder, **gram_op_kwargs)
+        self.normalize_by_num_channels = impl_params
+
+    def enc_to_repr(self, enc: torch.Tensor) -> torch.Tensor:
+        gram_matrix = super().enc_to_repr(enc)
+        if not self.normalize_by_num_channels:
+            return gram_matrix
+
+        num_channels = gram_matrix.size()[-1]
+        return gram_matrix / num_channels
+
+
+def ulyanov_et_al_2016_style_loss(
+    impl_params: bool = True,
+    instance_norm: bool = True,
+    stylization: bool = True,
+    multi_layer_encoder: Optional[MultiLayerEncoder] = None,
+    layers: Optional[Sequence[str]] = None,
+    layer_weights: Union[str, Sequence[float]] = "sum",
+    score_weight: float = None,
+    **gram_op_kwargs,
+):
+    if score_weight is None:
+        if impl_params:
+            if instance_norm:
+                score_weight = 1e0
+            else:
+                score_weight = 1e3 if stylization else 1e0
+        else:
+            score_weight = 1e0
+
+    if multi_layer_encoder is None:
+        multi_layer_encoder = ulyanov_et_al_2016_multi_layer_encoder()
+
+    if layers is None:
+        if impl_params:
+            layers = (
+                ("relu_1_1", "relu_2_1", "relu_3_1", "relu_4_1")
+                if instance_norm
+                else ("relu_1_1", "relu_2_1", "relu_3_1", "relu_4_1", "relu_5_1")
+            )
+        else:
+            layers = ("relu_1_1", "relu_2_1", "relu_3_1", "relu_4_1", "relu_5_1")
+
+    def get_encoding_op(encoder, layer_weight):
+        return UlyanovEtAl2016GramOperator(
+            encoder, score_weight=layer_weight, **gram_op_kwargs
+        )
+
+    return MultiLayerEncodingOperator(
+        multi_layer_encoder,
+        layers,
+        get_encoding_op,
+        layer_weights=layer_weights,
+        score_weight=score_weight,
+    )
+
+
+class UlyanovEtAl2016PerceptualLoss(MultiOperatorLoss):
+    def __init__(
+        self,
+        style_loss: MultiLayerEncodingOperator,
+        content_loss: MSEEncodingOperator = None,
+        stylization: bool = True,
+    ) -> None:
+
+        dict = OrderedDict([("style_loss", style_loss)])
+        if stylization:
+            dict["content_loss"] = content_loss
+
+        super().__init__(dict)
+
+    def set_content_image(self, image: torch.Tensor):
+        self.content_loss.set_target_image(image)
+
+    def set_style_image(self, image: torch.Tensor):
+        self.style_loss.set_target_image(image)
+
+    def forward(self, input_image: torch.Tensor) -> pystiche.LossDict:
+        for encoder in self._multi_layer_encoders:
+            encoder.encode(input_image)
+
+        loss = pystiche.LossDict(
+            [(name, op(input_image)) for name, op in self.named_children()]
+        )
+
+        for key in loss.keys():
+            loss[key] = loss[key] / input_image.size()[0]
```

Comment:
What are you doing here?

Reply:
The loss is divided here by the batch_size, just like the author did in his programming.

Reply:
The `MSEEncodingOperator` and the `GramOperator` internally use [`mse_loss`](https://github.com/pytorch/pytorch/blob/36e36eff2f8d1f8b695d20c89f27d3f27867e6c5/torch/nn/functional.py#L2516) from `torch.nn.functional`. The default loss `reduction` is `"mean"` and I think this includes averaging over all batches. Could you check that please?

Reply:
Yes, the default loss `reduction` is `"mean"` and this includes averaging over all batches. But the author also uses the default settings of  `nn.MSECriterion` in his implementation and then divides again by the `batch_size` before the error is backpropagated.
[https://github.com/torch/nn/blob/master/doc/criterion.md#nn.MSECriterion ](url)

Reply:
The link is not working, please fix it. The URL that is visible is different from the linked one. The syntax should be `[text that is shown](url that is linked)`. In your case that could be 

[nn.MSECriterion](https://github.com/torch/nn/blob/master/doc/criterion.md#nn.MSECriterion)

---

In that case I would create custom `Operator`s that implement this functionality in `calculate_score`. Check out the [`GatysEckerBethge2015MSEEncodingOperator`](https://github.com/pmeier/pystiche/blob/f5d42058af399b5ba162c439d7c6c42396315e9f/pystiche/papers/gatys_ecker_bethge_2015/loss.py#L27-L38) for an example how to do that.